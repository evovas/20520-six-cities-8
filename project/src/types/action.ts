import {Offer} from './offers';

export enum ActionType {
  SelectCity = 'booking/selectCity',
  ProcessOffers = 'data/processOffers',
}

export type SelectCityAction = {
  type: ActionType.SelectCity;
  payload: string;
}

export type ProcessOffersAction = {
  type: ActionType.ProcessOffers;
  payload: Offer[];
}

export type Actions = SelectCityAction | ProcessOffersAction;
