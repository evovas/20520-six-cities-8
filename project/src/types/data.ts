export type User = {
  id: number;
  name: string;
  avatarUrl: string;
  isPro: boolean;
}

export type ServerUser = {
  id: number;
  name: string;
  'avatar_url': string;
  'is_pro': boolean;
}

export type Location = {
  latitude: number;
  longitude: number;
  zoom: number;
}

export type City = {
  location: Location;
  name: string;
}

export type Offer = {
  id: number;
  bedrooms: number;
  city: City;
  description: string;
  goods: string[];
  host: User;
  images: string[];
  isFavorite: boolean;
  isPremium: boolean;
  location: Location;
  maxAdults: number;
  previewImage: string;
  price: number;
  rating: number;
  title: string;
  type: string;
}

export type ServerOffer = {
  id: number;
  bedrooms: number;
  city: City;
  description: string;
  goods: string[];
  host: ServerUser;
  images: string[];
  'is_favorite': boolean;
  'is_premium': boolean;
  location: Location;
  'max_adults': number;
  'preview_image': string;
  price: number;
  rating: number;
  title: string;
  type: string;
}

export type Review = {
  id: number;
  comment: string;
  date: Date;
  rating: number;
  user: User;
}

export type ServerReview = {
  id: number;
  comment: string;
  date: Date;
  rating: number;
  user: ServerUser;
}
