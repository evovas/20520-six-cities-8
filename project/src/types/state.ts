import {Offer} from './offers';

export type State = {
  currentCityName: string;
  offers: Offer[] | [];
}
