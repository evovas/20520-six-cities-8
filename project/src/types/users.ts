export type User = {
  id: number;
  name: string;
  avatarUrl: string;
  isPro: boolean;
}

export type Users = User[];
