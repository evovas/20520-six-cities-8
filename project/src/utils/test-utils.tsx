import {Action, AnyAction, createStore, Store} from '@reduxjs/toolkit';
import {ReactNode} from 'react';
import {render, RenderResult} from '@testing-library/react';
import {rootReducer} from '../store/root-reducer';
import {Provider} from 'react-redux';
import {Router} from 'react-router-dom';
import {createMemoryHistory, MemoryHistory} from 'history';

interface RenderWithRedux<
  S = any,
  A extends Action = AnyAction,
  I extends S = any
  > {
  (
    ui: ReactNode,
    reduxOption?: {
      preloadedState?: I
      store?: Store<S, A>
    },
    memoryHistory?: {
      history?: MemoryHistory,
    }
  ) : RenderResult
}

const renderWithRedux: RenderWithRedux = (
  ui: ReactNode,
  {
    preloadedState,
    store = createStore(rootReducer, preloadedState),
  } = {},
  {
    history,
  } = {},
) => ({
  ...render(
    <Provider store={store}>
      <Router history={history ?? createMemoryHistory()}>
        {ui}
      </Router>
    </Provider>,
  ),
});

export * from '@testing-library/react';
export { renderWithRedux };
