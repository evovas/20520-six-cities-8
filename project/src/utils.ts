import {Offer} from './types/data';

export const calculateRatingStars = (rating: number): string => {
  const STARS_COUNT = 5;
  return `${Math.round(rating) * 100 / STARS_COUNT}%`;
};

export const spliceOffersByCity = (offers: Offer[]): {[key: string]: Offer[]} => (
  offers.reduce<{[city: string]: Offer[]}>((acc, offer)=> {
    const city = offer.city.name;
    if (!acc[city]) {
      acc[city] = [];
    }
    acc[city].push(offer);
    return acc;
  }, {})
);

export const getRandomPositiveInteger = (a: number, b:number): number => {
  const lower = Math.ceil(Math.min(Math.abs(a), Math.abs(b)));
  const upper = Math.floor(Math.max(Math.abs(a), Math.abs(b)));
  const result = Math.random() * (upper - lower + 1) + lower;

  return Math.floor(result);
};
