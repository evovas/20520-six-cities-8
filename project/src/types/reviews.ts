import {User} from './users';

export type Review = {
  id: number;
  comment: string;
  date: Date;
  rating: number;
  user: User;
}

export type Reviews = Review[];
