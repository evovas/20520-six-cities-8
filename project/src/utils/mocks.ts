import {commerce, image, datatype, helpers, name, address, date, internet} from 'faker';
import {City, CurrentUser, Location, Offer, Review, User} from '../types/data';
import {getRandomPositiveInteger} from '../utils';
import dayjs from 'dayjs';

export const makeFakeUser = (): User => ({
  id: getRandomPositiveInteger(1, 100000),
  name: name.firstName(),
  avatarUrl: image.imageUrl(),
  isPro: datatype.boolean(),
});

export const makeFakeCurrentUser = (): CurrentUser => ({
  id: getRandomPositiveInteger(1, 100000),
  avatarUrl: image.imageUrl(),
  email: internet.email(),
  isPro: datatype.boolean(),
  name: name.firstName(),
  token: datatype.string(25),
});

export const makeFakeLocation = (): Location => ({
  latitude: +address.latitude(),
  longitude: +address.longitude(),
  zoom: getRandomPositiveInteger(8, 12),
});

export const makeFakeCity = (): City => ({
  location: makeFakeLocation(),
  name: address.city(),
});

export const makeFakeOffer = (): Offer => ({
  id: getRandomPositiveInteger(1, 10000000),
  bedrooms: getRandomPositiveInteger(1, 6),
  city: makeFakeCity(),
  description: commerce.productDescription(),
  goods: [commerce.productMaterial(), commerce.productMaterial(), commerce.productMaterial()],
  host: makeFakeUser(),
  images: [
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
    image.imageUrl(),
  ],
  isFavorite: datatype.boolean(),
  isPremium: datatype.boolean(),
  location: makeFakeLocation(),
  maxAdults: getRandomPositiveInteger(1, 6),
  previewImage: image.imageUrl(),
  price: getRandomPositiveInteger(100, 1000),
  rating: getRandomPositiveInteger(10, 50) / 10,
  title: commerce.productName(),
  type: helpers.randomize(['apartment', 'room', 'house', 'hotel']),
});

export const makeFakeReview = (): Review => ({
  id: getRandomPositiveInteger(1, 10000000),
  comment: datatype.string(getRandomPositiveInteger(50, 300)),
  date: dayjs(date.past()).toISOString(),
  rating: getRandomPositiveInteger(1, 5),
  user: makeFakeUser(),
});
