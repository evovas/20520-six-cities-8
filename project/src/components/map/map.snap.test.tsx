import Map from './map';
import {renderWithRedux} from '../../utils/test-utils';
import {City, Offer} from '../../types/data';

const fakeOffers: Offer[] = [];
const fakeCity: City = {
  location: {
    latitude: 52.370216,
    longitude: 4.895168,
    zoom: 10,
  },
  name: 'Paris',
};

describe('Component: Map', () => {
  test('should render correctly', () => {
    const {container} = renderWithRedux(
      <Map
        className={'cities__map'}
        offers={fakeOffers}
        city={fakeCity}
        activeCardId={null}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
