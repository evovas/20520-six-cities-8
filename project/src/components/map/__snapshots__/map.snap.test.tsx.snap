// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Map should render correctly 1`] = `
<div>
  <section
    class="map cities__map leaflet-container leaflet-grab leaflet-touch-drag"
    data-testid="map"
    style="position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      >
        <div
          class="leaflet-layer "
          style="z-index: 1;"
        >
          <div
            class="leaflet-tile-container leaflet-zoom-animated"
            style="z-index: 18; left: 0px; top: 0px;"
          >
            <img
              alt=""
              class="leaflet-tile"
              role="presentation"
              src="https://a.basemaps.cartocdn.com/rastertiles/voyager/10/525/336.png"
              style="width: 256px; height: 256px; left: -237px; top: -134px;"
            />
          </div>
        </div>
      </div>
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            −
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
           | © 
          <a
            href="https://www.openstreetmap.org/copyright"
          >
            OpenStreetMap
          </a>
           contributors © 
          <a
            href="https://carto.com/attributions"
          >
            CARTO
          </a>
        </div>
      </div>
    </div>
  </section>
</div>
`;
